### DISCUSSION TITLE
"Value error, Expected UploadFile, received: <class 'str'>" When uploading file without filename.

### DISCUSSION THREAD

**Discussion by 4c0n** [fastapi/fastapi#11822] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI

app = FastAPI()


@app.post("/upload")
def upload(file: UploadFile):
    return {"file", file.filename}
```

### Description

When we use httpx to upload a file without a filename, FastAPI seems to give a unexpected response.
Now I realize that this is not a very common case, however it is possible and correct to upload a file omitting the name in the multipart form post request.

To reproduce consider the following code:
```python
import httpx


async def test_upload() -> None:
    async with httpx.AsyncClient() as client:
        response = await client.post(
            "http://localhost:8000/upload",
            files={"file": (None, open("test_file.txt", "rb")},
            headers={"Content-Type": "multipart/form-data; boundary=----FILEDATA"},
        )
        print(response.text)
```
Unfortunately this results in the 422 response:

```
{"detail":[{"type":"value_error","loc":["body","file"],"msg":"Value error, Expected UploadFile, received: <class 'str'>","input":"Hello World!","ctx":{"error":{}},"url":"https://errors.pydantic.dev/2.8/v/value_error"}]}
```

Where one would expect a 200 response with body:
```
{"file": null}
```
as the `filename` property of `UploadFile` is `None` by default.

Please correct me if I am wrong, but it seems like an edge case that FastAPI currently does not support.
It is a bit annoying, because in order to satisfy mypy we have to check that the `filename` is not `None` if we want to use it in our code, however we cannot currently test the case where it actually is `None`, because the input validation fails in a odd way.

Not sure where to look when it comes to fixing this, but I would be willing to spend some time debugging this if someone can point me in the right direction.

### Operating System

Linux

### FastAPI Version

0.109.2

### Pydantic Version

2.8.2

### Python Version

3.12.4

**Comment by jgould22** [comment, marked as answer]
> however it is possible and correct to upload a file omitting the name in the multipart form post request.

In starlette's case it is not.  You can read more here https://www.starlette.io/requests/#Request%20Files

> As settled in [RFC-7578: 4.2](https://www.ietf.org/rfc/rfc7578.txt), form-data content part that contains file assumed to have name and filename fields in Content-Disposition header: Content-Disposition: form-data; name="user"; filename="somefile". Though filename field is optional according to RFC-7578, it helps Starlette to differentiate which data should be treated as file. If filename field was supplied, UploadFile object will be created to access underlying file, otherwise form-data part will be parsed and available as a raw string.

So this is more of a starlette thing than a FastAPI thing.

**Reply by 4c0n** [reply]
Even then, FastAPI does not handle it correctly do you not agree? This validation error states it is expecting an instance of UploadFile and is receiving a string. Which is weird, but apparently can be expected.

**Reply by jgould22** [reply]
I mean, sure but that is because by not including the filename starlette has assumed it is supposed to be a str and you have told fastapi that you wanted a file.

There is not much for fastapi to do here because this choice / processing is done in starlette before the value reaches fastapi's validation code.

**Reply by 4c0n** [reply]
Sure that sounds reasonable, but then why would the filename property of the FastAPI version of UploadFile need to be None by default, considering it will always be set?
I don't really need to support uploads without a name, but am currently stuck with untestable code and a supposedly completely useless None check of the filename property.

IMO we could make the property not nullable, but that would still produce this weird validation error when an upload without filename is received. 
Or we could, for example, raise a proper exception in this case stating that it is not supported to upload files without a name.

Another option would be to actually accept the string provided by Starlette and make it available through the UploadFile somehow.

Sure there's also the option to accept it for what it is, but I fail to see how that's very helpful for developers.

**Comment by YStrauchP4** [comment]
I am in a real pickle here.
I have a client library (`curl` for R) that does not allow adding a file name for in-memory uploads. Technically they are correct as RFC-7578 says for this use case, file names are allowed to be omitted. There should not be a need to add a name from their perspective.

From what I understand, Scarlette is deliberately implementing RFC-7578 "incorrectly" and relies on the name to determine it's a file upload. But they don't have an annotation that tells them to expect a file, so they have to guess it somehow.

FastAPI knows that I expect a file because I annotated the parameter. It doesn't have a mechanism to communicate this to Scarlette, Scarlette guesses the type incorrectly, and it gets rejected by FastAPI (pydantic probably).

I am not deep in the design patterns between FastAPI and Scarlette, but to me it seems that FastAPI needs an API to tell Scarlette to expect a file. Is that feasible?
I have opened a discussion on the Scarlette repo:
https://github.com/encode/starlette/discussions/2707

**Reply by rls-p4** [reply]
See related issue/discussion in the `curl` for R repo: https://github.com/jeroen/curl/issues/339
